pkg/operator/certrotation/metrics/monitor.go

```diff
@@ -0,0 +1,187 @@
+
+	timeNowFn = time.Now
+)
+
+var (
+	caBundleExpireHoursDesc = prometheus.NewDesc(
+		"certificates_ca_bundle_expire_hours",
+		"Number of hours until certificates in given CA bundle expire",
+		[]string{"namespace", "name", "common_name", "signer_name", "valid_from"}, nil)
+
+	signerExpireHoursDesc = prometheus.NewDesc(
+		"certificates_signer_expire_hours",
+		"Number of hours until certificates in given signer expire",
+		[]string{"namespace", "name", "common_name", "signer_name", "valid_from"}, nil)
+
+	targetExpireHoursDesc = prometheus.NewDesc(
+		"certificates_target_expire_hours",
+		"Number of hours until certificates in given target expire",
+		[]string{"namespace", "name", "common_name", "signer_name", "valid_from"}, nil)
+)
+
+type certMetricsCollector struct {
+	configLister corev1listers.ConfigMapLister
+	secretLister corev1listers.SecretLister
+
+	nowFn func() time.Time
+}
+
+// Describe implements the prometheus collector interface
+func (c *certMetricsCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- caBundleExpireHoursDesc
+	ch <- signerExpireHoursDesc
+	ch <- targetExpireHoursDesc
+}
+
+// Collect implements the prometheus collector interface
+func (c *certMetricsCollector) Collect(ch chan<- prometheus.Metric) {
+	// to speed up collection, do this in parallel
+	var wg sync.WaitGroup
+	wg.Add(2)
+	go func() { defer wg.Done(); c.collectCABundles(ch) }()
+	go func() { defer wg.Done(); c.collectSignersAndTarget(ch) }()
+	wg.Wait()
+}
+
+func (c *certMetricsCollector) collectSignersAndTarget(ch chan<- prometheus.Metric) {
+	secrets, err := c.secretLister.List(getCertificateManagedLabelSelector())
+	if err != nil {
+		klog.Warningf("Failed to list signer secrets: %v", err)
+		return
+	}
+
+	for _, secret := range secrets {
+		var targetDescType *prometheus.Desc
+
+		switch GetCertificateTypeFromObject(secret) {
+		case CertificateTypeSigner:
+			targetDescType = signerExpireHoursDesc
+		case CertificateTypeTarget:
+			targetDescType = targetExpireHoursDesc
+		default:
+			klog.Warningf("Secret %s/%s has unknown certificate type: %q", secret.Namespace, secret.Name, secret.Labels[CertificateManagedTypeLabelName])
+			continue
+		}
+		if secret.Data["tls.crt"] == nil || secret.Data["tls.key"] == nil {
+			klog.V(4).Infof("Secret %s/%s does not have 'tls.crt' or 'tls.key'", secret.Namespace, secret.Name)
+			continue
+		}
+
+		signingCertKeyPair, err := crypto.GetCAFromBytes(secret.Data["tls.crt"], secret.Data["tls.key"])
+		if err != nil {
+			continue
+		}
+
+		labelValues := []string{}
+		for _, certificate := range signingCertKeyPair.Config.Certs {
+			expireHours := certificate.NotAfter.UTC().Sub(c.nowFn().UTC()).Hours()
+			labelValues = append(labelValues, []string{
+				secret.Namespace,
+				secret.Name,
+				certificate.Subject.CommonName,
+				certificate.Issuer.CommonName,
+				fmt.Sprintf("%s", certificate.NotBefore.UTC()),
+			}...)
+
+			ch <- prometheus.MustNewConstMetric(
```

Comment:
I don't actually know what this is doing.  You sure this isn't a loop?

Reply:
we do the same in deployment configs metrics, we publishing single cert to prometheus